Update developer README (#20)
movestore/Template_R_Function_App: developer_README.md

# MoveApps R Software Development Kit (SDK)

***NOTE*: this SDK supports code written for input data of class `move2` and not `moveStack`, as all input data of class `moveStack` will be converted to class `move2`. For all other input/output types, this SDK works as usual. Please contact us under support@moveapps.org if you have any questions.**

This documentation provides a short introduction to the [MoveApps](https://www.moveapps.org) **R SDK**.

> As a first step, and before your read this, you should have used this GitHub template to create a copy of it in your personal space and named the repository as your App will be named in MoveApps (please adhere to our convention of Title Case, e.g. My New App).

Reviewer: ```suggestion
As a first step, and before your read this, you should have used this GitHub template to create a copy of it in your personal space and named the repository as your App will be named in MoveApps.
```

**The [MoveApps User Manual](https://docs.moveapps.org/#/create_app) provides a step-by-step explanation of how to create an App.** Please carefully follow these steps when creating a MoveApps App.

## Files in the SDK/template

This template is designed according to a file structure that is necessary for your App to run in your local development environment similar to the way it will run in the MoveApps environment later. Please contain the structure and only change/add files as necessary for your App's functionality. Take a look at the [overview in the User Manual](https://docs.moveapps.org/#/create_app) to see which files can be changed and which should remain as is for simulation of the behaviour on MoveApps on your local system.

> Here you find an overview of the files and their function in the SDK.

Reviewer: ```suggestion
Here you find an overview of the files and their function in the SDK:
```

> 1. `./RFunction.R`: This is the entrypoint for your App logic. MoveApps will call this function during a Workflow execution which includes your App. The file must be named `RFunction.R`, do not alter it. See [Step 3](https://docs.moveapps.org/#/create_app#step-3-develop-the-app-code-locally-within-the-template) in the User Manual.

Reviewer: ```suggestion
1. `./RFunction.R`: This is the entry point for your App logic. MoveApps will call this function during a Workflow execution which includes your App. The file must be named `RFunction.R`, do not alter it. See [Step 3](https://docs.moveapps.org/#/create_app#step-3-develop-the-app-code-locally-within-the-template) in the User Manual.
```

> 1. `./RFunction.R`: This is the entrypoint for your App logic. MoveApps will call this function during a Workflow execution which includes your App. The file must be named `RFunction.R`, do not alter it. See [Step 3](https://docs.moveapps.org/#/create_app#step-3-develop-the-app-code-locally-within-the-template) in the User Manual.

Reviewer: ```suggestion
1. `./RFunction.R`: must be modified by the developer. This is the entrypoint for your App logic. MoveApps will call this function during a Workflow execution which includes your App. The file must be named `RFunction.R`, do not alter it. See [Step 3](https://docs.moveapps.org/#/create_app#step-3-develop-the-app-code-locally-within-the-template) in the User Manual.
```

> 1. `./appspec.json`: This file defines the settings and metadata of your App. See [Step 5](https://docs.moveapps.org/#/create_app#step-5-write-app-specifications) in the User Manual.

Reviewer: ```suggestion
1. `./appspec.json`: must be modified by the developer. This file defines the settings and metadata of your App. See [Step 5](https://docs.moveapps.org/#/create_app#step-5-write-app-specifications) in the User Manual.
```

1. `./renv.lock`: Definition of the dependencies of your App. We use `renv` as library manager. Optional, see below.

> 1. `./data/**`: Resources of the SDK

Reviewer: ```suggestion
1. `./sdk.R`: used for App testing. The main entry point of the SDK. Use it to execute your App in your compiler (e.g. RStudio).
1. `./app-configuration.json`: used for App testing. The configuration/settings file of your App (in [JSON](https://www.w3schools.com/js/js_json_intro.asp) format - must correspondent with the `settings` of your `appspec.json`, see [MoveApps parameters](https://docs.moveapps.org/#/copilot-r-sdk.md#moveapps-parameters) for an example of the `app-configuration.json` file).
1.  `/.env`: defining the SDK Runtime environment, see below. Make sure to check "Show Hidden Files" in the settings menu of the "Files" tab in RStudio.
1. `./data/**`: used for App testing. Resources of the SDK
```

   1. `auxiliary/**`: Simulates the usage of [*auxiliary files*](https://docs.moveapps.org/#/auxiliary). You can put files into this folder to simulate an App run with provided/user-uploaded files. 

>    1. `output/**`: If your App produces output files (artifacts) they will be stored here. See [*producing artifacts*](https://docs.moveapps.org/#/copilot-r-sdk?id=producing-artefacts) for more information.

Reviewer: ```suggestion
   1. `output/**`: The output data (`output.rds`) that will be passed on to the next App in a Workflow and other output files (artifacts) that your App may produce will be stored here. See [*producing artifacts*](https://docs.moveapps.org/#/copilot-r-sdk?id=producing-artefacts) for more information.
```

>    1. `raw/**`: Collection of sample App input data. You can use these samples to simulate an App run with real input.

Reviewer: ```suggestion
   1. `raw/**`: Collection of sample App input data. You can use these data to test an App run with real input.
```

> 1. `./sdk/**`: The (internal) MoveApps R SDK logic.

Reviewer: ```suggestion
1. `./src/**`: The (internal) MoveApps R SDK logic.
```

> 1. `./sdk.R`: The main entry point of the SDK. Use it to execute your App in your compiler (e.g. RStudio).

Reviewer: ```suggestion
```

> 1. `./tests/**`: Location for *Unit Tests*. See [Step 4](https://docs.moveapps.org/#/create_app#=step-4-test-your-app-locally) of the User Manual.

Reviewer: ```suggestion

```

1. `.env`: defining the SDK Runtime environment, see below.

## SDK runtime environment

> Critical parts of the SDK can be adjusted by `environment variables`. Keep in mind that these variables are only changeable during App development and not during an App run on MoveApps. They are predefined with sensible defaults - they should work for you as they are.

Reviewer: ```suggestion
Critical parts of the SDK can be adjusted by `environment variables`. Keep in mind that these variables are only changeable during App development and not during an App run on MoveApps. They are predefined with sensible defaults - they should work for you as they are. The only variable that you will want to adjust is `SOURCE_FILE` to call the different example data provided in the template or other data that you want to test your App with.
```

Note that many App features will be set and updated with information from the `appspec.json` in each new App version. Thus, even if not bundled into the App, this file is required and must be up to date.

> You are also required to write a documentation file (see [Step 6](https://docs.moveapps.org/#/create_app#step-6-write-a-documentation-file) in the User Manual).

Reviewer: ```suggestion

```

## R packages management / renv (optional)

The template is prepared to use [`renv` as a dependency manager](https://rstudio.github.io/renv/articles/renv.html) - but is disabled by default (_opt-in_).
You can [activate `renv` with `renv::activate()`](https://rstudio.github.io/renv/articles/renv.html#uninstalling-renv) and then use it in the [usual `renv` workflow](https://rstudio.github.io/renv/articles/renv.html#workflow).

### Docker support (optional)

> - at the end your app will be executed on MoveApps in a Docker container.

Reviewer: ```suggestion
- Your app will be executed on MoveApps in a Docker container. Specially for debugging errors that are not straight forward, it can be very useful to execute your App in a docker environment. These more complex error are often due to system libraries that need to be installed in MoveApps to run the App. The easiest way to find out which ones they are is to run the App in a docker environment.
```
